[boinkkitty] iP

boinkkitty opened the pull request:
# Boink

>> "Your mind is for having ideas, not holding them." - David Allen([source](https://dansilvestre.com/productivity-quotes/))

Boink frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2425s2.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [x] Managing tasks
- [] Managing deadlines (coming soon)
- [] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename, Greet, Exit

Commit:
Echo

Commit:
Add, List

Commit:
Mark as Done

Commit:
ToDo, Deadline, Event

Commit:
Automated Text UI Testing

Commit:
Exception Handling

Commit:
Delete function

Commit:
Add TaskList.java with load, save and task methods

Commit:
Add documentation for Task subclass methods

bipplane commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+        return "OK, I've removed this task from the list \n" + removedTask;
+    }
+
+    public String printTasks() {
+        List<String> output = IntStream.range(0, this.tasks.size()).mapToObj(
+                i -> ((i + 1) + "." + this.tasks.get(i) + "\n"))
+                .collect(Collectors.toList());
+        return String.join("", output);
```
This implementation is quite intuitive!

wowwwp replied:
I agree, this is a very smart way of printing the tasks

bipplane commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+        return String.join("", output);
+    }
+
+    public void saveTasks() {
+        try {
+            FileWriter fw = new FileWriter("../../../data/data.txt", false);
+            StringBuilder sb = new StringBuilder();
+            for (Task task: tasks) {
+                sb.append(task.saveTask() + "\n");
+            }
+            fw.write(sb.toString());
+            fw.close();
+        } catch (IOException e) {
+            System.out.println("Error occurred while saving tasks to file " + e.getMessage());
+        }
+    }
```
remember to mention that this method throws IOException

bipplane commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+     * Throws an exception if file does not exist.
+     * @param filePath The location of file to be read from
+     */
+    public void loadTasks(String filePath) throws FileNotFoundException {
+        File tasksFile = new File(filePath);
+        Scanner scanner = new Scanner(tasksFile);
+
+        while (scanner.hasNext()) {
+            String line = scanner.nextLine();
+            String[] parts = line.split("\\|");
+
+            if (parts[0].trim().equals("T")) {
+                ToDo todo = new ToDo(parts[2].trim());
+                if (parts[1].trim().equals("1")) {
+                    todo.setDone();
+                }
+                this.tasks.add(todo);
+            } else if (parts[0].trim().equals("D")) {
+                Deadline deadline = new Deadline(parts[2].trim(), parts[3].trim());
+                if (parts[1].trim().equals("1")) {
+                    deadline.setDone();
+                }
+                this.tasks.add(deadline);
+            } else if (parts[0].trim().equals("E")) {
+                Event event = new Event(parts[2].trim(), parts[3].trim(), parts[4].trim());
+                if (parts[1].trim().equals("1")) {
+                    event.setDone();
+                }
+                this.tasks.add(event);
+            }
+        }
+    }
```
Remember to write a catch-block for `FileNotFoundException` in this method!

bipplane commented on src/main/java/ToDo.java:
```diff
@@ -3,6 +3,12 @@ public ToDo (String name) {
         super(name);
```
Consider renaming this class to something such as `ToDoTask` to reflect its inheritance from `Task`, making it easier to identify.

bipplane commented on src/main/java/BoinkChatter.java:
```diff
@@ -1,12 +1,11 @@
     public static void main(String[] args) {
         Scanner scanner = new Scanner(System.in);
-        ArrayList<Task> tasks = new ArrayList<>();
+        TaskList taskList = new TaskList();
```
Good use of `TaskList` to manage tasks more effectively than a plain `ArrayList<Task>`

bipplane commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * Constructor for task list
+     */
+    public TaskList() {
+        try {
+            this.loadTasks("../../../data/data.txt");
+        } catch (FileNotFoundException e) {
+            System.out.println("Error loading tasks! " + e.getMessage());
+        }
+    }
```
This method should be indicated to be throwing `FileNotFoundException`

bipplane reviewed (commented):
Great job on the iP so far! ❤️

Commit:
Add documentation for TaskList methods

Commit:
Rename ToDo class to ToDoTask

Commit:
Add parsing of date and time as objects

wowwwp commented on src/main/java/BoinkChatter.java:
```diff
@@ -0,0 +1,102 @@
+        Scanner scanner = new Scanner(System.in);
+        TaskList taskList = new TaskList();
+
+        String logo = "Hello! I'm BoinkChatter\n" +
```
I like how you split the long line into 2, it makes it more readable

wowwwp commented on src/main/java/BoinkChatter.java:
```diff
@@ -0,0 +1,102 @@
+     * Throws DateTimeParseException if input format is incorrect.
+     * @param input String datetime (dd/MM/yyyy HHmm)
+     * @return LocalDateTime object
+     * @throws DateTimeParseException
```
Perhaps it is better to state the reason why the error gets thrown next to the `@throws DateTimeParseException` ?

wowwwp commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /**
+     * Converts LocalDateTime to String output in (dd MMM yyyy HH:mm) format
```
I like how detailed the JavaDoc comment is

wowwwp commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,68 @@
+     * @return DateTime (dd MMM yyyy HH:mm).
+     */
+
+    public String getDateTime(LocalDateTime dt) {
```
Should the variable name be more detailed?

wowwwp commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+
+            if (parts[0].trim().equals("T")) {
+                ToDoTask todo = new ToDoTask(parts[2].trim());
+                if (parts[1].trim().equals("1")) {
```
Perhaps it is better to give `parts[1]` a meaningful name to improve readability?

wowwwp reviewed (commented):
Looks good. Just a few variable names which can be changed.

Commit:
Refactor and extract classes to Storage, Ui, Parser, Command

DanielJames0302 commented on src/main/java/boink/Command.java:
```diff
@@ -0,0 +1,9 @@
+package boink;
+
+public abstract class Command {
+    public abstract void execute(Storage storage, Ui ui, TaskList tasks);
```
I like this `execute()` function which allows to specify how task can be executed

DanielJames0302 commented on src/main/java/boink/Parser.java:
```diff
@@ -0,0 +1,88 @@
+            }
+        default:
+            return new UnknownCommand();
+        }
```
Well aware of switch-case indentation

DanielJames0302 commented on src/main/java/boink/Task.java:
```diff
@@ -0,0 +1,70 @@
+        return this.name;
+    }
+
+    public boolean checkDone() {
```
Perhaps a more boolean name for this method ?

DanielJames0302 reviewed (commented):
LGTM. Just some nits to fix.

Commit:
Merge branch 'add-gradle-support' of https://github.com/boinkkitty/ip

Commit:
Add Gradle support

Commit:
Add JUnit Testing for Parser and TaskList

Commit:
Package App as JAR file

Commit:
Add JavaDoc for main classes

Commit:
Tweak code to comply with coding standard

Commit:
Merge branch 'branch-CodingStandard'

Commit:
Add find command to tasklist

Commit:
Implement checkStyle and rectify coding style violations

Commit:
Add GUI to Boink Chatbot

Commit:
refactor: Replace Command class with Enum

Refactored the Command class into an Enum to simplify command management
and improve type safety. This change removes unnecessary boilerplate code
and makes it easier to add new commands in future.

- Eliminated the Command class and replaced it with an Enum.
- Each command is now defined as an Enum constant with associated behavior.

This refactor does not change existing functionality but improves code
structure and readability.

Commit:
Add exception handling, Utils class and update testcases

InvalidIndexException requires handling to prevent sudden crash
during program execution.

Utils class contain various utility methods to streamline code
and improve reusability.

Testcases updated to ensure method tests work as intended.

Commit:
Add assertion for improved validation

Commit:
Add use of stream in tasklist

Commit:
Improve code quality and apply SLAP

boinkkitty committed:
Merge pull request #2 from boinkkitty/branch-A-Assertions

Add assertion for improved validation

boinkkitty committed:
Merge branch 'master' into branch-A-CodeQuality

boinkkitty committed:
Merge pull request #3 from boinkkitty/branch-A-CodeQuality

Improve code quality and apply SLAP

boinkkitty committed:
Merge pull request #4 from boinkkitty/branch-A-Streams

Add use of stream in tasklist

Commit:
Add Archive command to Boink

boinkkitty committed:
Merge pull request #5 from boinkkitty/branch-C-Archive

Add Archive command to Boink

Commit:
Modify testcase and Boink picture

Commit:
Modify hardcoded filepath to relative filepath

boinkkitty committed:
Merge pull request #6 from boinkkitty/branch-A-AiAssisted

Modify testcase and Boink picture

Commit:
Update README, add Ui.png and error handling

boinkkitty committed:
Merge pull request #7 from boinkkitty/branch-A-UserGuide

Update README, add Ui.png and error handling
